src/main/java/org/truffleruby/core/thread/ThreadNodes.java

```diff
@@ -376,6 +381,47 @@ public DynamicObject pass() {
 
     }
 
+    @CoreMethod(names = "safe_level")
+    public abstract static class SafeLevelNode extends CoreMethodArrayArgumentsNode {
+
+        @Child private ReadObjectFieldNode readNode;
+
+        @Specialization
+        public int safeLevel(DynamicObject self) {
+            return (int)getReadNode().execute(Layouts.THREAD.getThreadLocals(self));
+        }
+
+        private ReadObjectFieldNode getReadNode() {
+            if (readNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                readNode = insert(ReadObjectFieldNodeGen.create("$SAFE", nil()));
+            }
+            return readNode;
+        }
+
+    }
+
+    @Primitive(name = "thread_set_safe_level_force", lowerFixnum = 1)
+    public static abstract class ThreadSetSafeLevelForceNode extends PrimitiveArrayArgumentsNode {
+
+        @Child private WriteObjectFieldNode writeSafeLevel;
+
+        @Specialization(guards = "isRubyThread(thread)")
+        public int safeLevel(DynamicObject thread, int safeLevel) {
+            writeSafeLevel(Layouts.THREAD.getThreadLocals(thread), safeLevel);
```

Comment:
Can we reuse a GetThreadLocalsObjectNode?

Reply:
Yes, fixed.